### ISSUE TITLE
Improving performance of Node.connect

### ISSUE THREAD

**Issue by pmcharrison** [Dallinger/Dallinger#2326]
`Node.connect` checks to make sure that the node isn't already connected to the new node:

```
new_vectors = []
        if direction in ["to", "both"]:
            already_connected_to = self.flatten(
                [self.is_connected(direction="to", whom=whom)]
            )
            for node, connected in zip(whom, already_connected_to):
                if connected:
                    print(
                        "Warning! {} already connected to {}, "
                        "instruction to connect will be ignored.".format(self, node)
                    )
                else:
                    new_vectors.append(Vector(origin=self, destination=node))
```

This is slow because `is_connected` requires a database query:

```
vectors = (
                Vector.query.with_entities(Vector.destination_id)
                .filter_by(origin_id=self.id, failed=False)
                .all()
            )
```

I would propose removing this check, because it makes things much slower for bulk node creation.

**Comment by thomasmorgan** [comment]
Has this inefficiency caused problems? I would be concerned about the possibility of nodes having multiple vectors joining them which would cause problems.

That said, if an experiment used different subclasses of vector this could become a feature.

**Comment by pmcharrison** [comment]
> Has this inefficiency caused problems? 

It caused problems for us, yes (experiments failing to start). 

> I would be concerned about the possibility of nodes having multiple vectors joining them which would cause problems.

Agreed, this is why I think it makes sense to keep the safety check as a default, only disabled when the researcher explicitly requests it.

**Comment by thomasmorgan** [comment]
How confident are you that this bit of code is responsible? I ask because that query is pretty light and the relevant columns are indexed, so unless you are creating many thousands of vectors all at once I'd be surprised if this was an issue. That said, I did run into issues with an experiment that created 200,000 vectors.

If it is definitely the issue there are a couple of other fixes: (1) Can `connect` or `is_connected` be further optimized? (2) Can the database better support large volumes of this query? (3) Your experiment could create a custom subclass of Node to allow this functionality without touching core dallinger, (4) Can you tweak your experiment to avoid excess vector creation (thats what we did in our case).

This is a good concern, I'm just cautious about even allowing a secret backdoor to duplicate vector creation.

**Comment by pmcharrison** [comment]
> How confident are you that this bit of code is responsible?

Yes, totally confident - I already resolved this issue on a local Dallinger branch by disabling the check and it fixed the problem.

> I ask because that query is pretty light and the relevant columns are indexed, so unless you are creating many thousands of vectors all at once I'd be surprised if this was an issue. That said, I did run into issues with an experiment that created 200,000 vectors.

Yes, the issue comes when populating a large network (1000s) all in one go.

> If it is definitely the issue there are a couple of other fixes: (1) Can connect or is_connected be further optimized? (2) Can the database better support large volumes of this query? 

Yes, maybe the developers could think a little about these possibilities. The best alternative I can think of is to load the vectors eagerly as part of the original node query (see https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#joined-eager-loading-and-result-set-batching) with the goal that no extra SQL query is required to test `is_connected`, but I have not used this functionality before so don't know whether it would actually work here. Such a solution would require significant thought/testing and hence be expensive in terms of developer time, whereas my original proposed fix would just take a minute to implement.

> (3) Your experiment could create a custom subclass of Node to allow this functionality without touching core dallinger, 

Possible, but hijacking methods like this without calling `super()` is in itself dodgy because you lose access to future Dallinger changes to the underlying method.

> (4) Can you tweak your experiment to avoid excess vector creation (thats what we did in our case).

Possible, but much less desirable than overcoming this fundamental performance limitation.

> This is a good concern, I'm just cautious about even allowing a secret backdoor to duplicate vector creation.

I would argue that it's totally Pythonic to provide this. One of the core Python mantras is "We are all consenting adults here". The idea is that the programmer is given the license to bypass certain securities as long as one takes responsibility for managing the implications. I don't see how this change would negatively influence anyone, noone is forced to disable the checks.

**Comment by thomasmorgan** [comment]
OK, sounds good. Let's first look into further optimizations and, if that doesn't work out, go from there. The current code is likely quite unoptimized though (I wrote it!). One possibility is to replace this:

```
vectors = (
                Vector.query.with_entities(Vector.destination_id)
                .filter_by(origin_id=self.id, failed=False)
                .all()
            )
```
With this:
```
vectors = self.all_outgoing_vectors  # this is a relationship backref that semi-dodges a query and may be very fast
vectors = [v for v in vectors if !v.failed]  # simple python to filter out failed vectors
```

Another option you should consider is dodging `connect` entirely. The `node.connect` function is just a wrapper for vector creation that adds a bunch of checks. If you don't want these checks just make the vector directly:
e.g. instead of:
```
node1.connect(node2, direction="to")
```
just do:
```
Vector(origin=node1, destination=node2)
```
This is always going to be the fastest way to connect nodes.

**Comment by pmcharrison** [comment]
Thanks very much, @thomasmorgan! I didn't see `all_outgoing_vectors`, that's helpful. Unfortunately I think by default this is also lazily evaluated so would still require a fresh query.

However, I think I'm sold on your idea of just using `Vector` directly! I think it makes perfect sense here. So, I'd like to revise this request: instead of changing to the function itself, we just add a little note in the documentation suggesting using `Vector` directly in performance-sensitive applications.

**Comment by thomasmorgan** [comment]
This is good, yes. We should make clear that some functions are just wrappers on object creation that perform checks, and to skip these checks just directly create the objects:

node.connect() -> Vector()
node.transmit() -> Transmission()

Are these the only ones? The init function of these objects do a few minor checks too. It might be worth looking at them to see if they do costly queries too.

**Comment by pmcharrison** [comment]
Thanks, good point about transmit()! Yes, let's document that too.

Good point about the init functions too - I just had a look, I think
they're OK.
